modules/standard/Math.chpl

```diff
@@ -1232,4 +1232,80 @@ module Math {
   }
+
+  @chpldoc.nodoc
+  private inline proc fmaImplExternCall(x: real(?w), y: real(w), z: real(w)) {
+    extern proc fmaf(x: real(32), y: real(32), z: real(32)): real(32);
+    extern proc fma(x: real(64), y: real(64), z: real(64)): real(64);
+    var ret: x.type;
+    select x.type {
+      when real(32) do ret = fmaf(x, y, z);
+      when real(64) do ret = fma(x, y, z);
+      otherwise do compilerError('Unhandled type in fma() impl!');
+    }
+    return ret;
+  }
+
+  @chpldoc.nodoc
+  private inline
+  proc fmaSelectPrimitiveOrExternCall(x: real(?w), y: real(w), z: real(w)) {
+    param targetCompiler = __primitive("get compiler variable",
+                                       "CHPL_TARGET_COMPILER");
+    param isLlvmCompile = targetCompiler == "llvm";
+
+    // The backend will emit a 'llvm.fma.*' instruction, which should be
+    // optimized into a hardware instruction if the architecture is
+    // specified and `--specialize` is thrown.
+    if isLlvmCompile {
+      return __primitive("fma", x, y, z);
+
+    // Call C 'fma()' here rather than have the backend do it. It's up to
+    // the C compiler if any optimization occurs here at all. We have
+    // to call 'fma()' for correctness, as (x*y+z) will introduce more error.
+    // TODO: Is there a path for the builtin Clang to reliably emit a
+    // hardware instruction, e.g., through a Clang intrinsic call?
+    } else {
+      return fmaImplExternCall(x, y, z);
+    }
+  }
+
+  /*  Performs a fused multiply-add operation that multiplies ``x`` and ``y``
+      and adds ``z`` to the result. The advantage of ``fma()`` over the
+      expression ``(x*y)+z`` is that it avoids the additional error
+      introduced by performing two separate floating point operations.
+      It can also be faster on machines that implement the operation as a
+      single instruction.
+
+      .. note::
+
+        When compiling with ``CHPL_TARGET_COMPILER=llvm``, this procedure
+        should reliably generate a single hardware instruction on ``x86``
+        if ``--specialize`` is thrown and ``CHPL_TARGET_CPU`` is set
+        (provided that the ``x86`` CPU supports hardware FMA).
+
+        When compiling with C, this procedure will call out to the ``fma()``
+        routines defined in the C header `math.h`. Any optimization performed
+        is decided by the C compiler.
+  */
+  @unstable("The 'fma()' procedure was recently added, and may change based on feedback")
+  inline proc fma(x: real(64), y: real(64), z: real(64)): real(64) {
+    return fmaSelectPrimitiveOrExternCall(x, y, z);
+  }
+
+  @unstable("The 'fma()' procedure was recently added, and may change based on feedback")
+  inline proc fma(x: real(32), y: real(32), z: real(32)): real(32) {
+    return fmaSelectPrimitiveOrExternCall(x, y, z);
+  }
+
+  @unstable("The 'fma()' procedure was recently added, and may change based on feedback")
+  inline proc fma(x: imag(64), y: imag(64), z: imag(64)): imag(64) {
```

Comment:
@dlongnecke-cray — It was recently pointed out to me that, if `x`, `y`, and `z` are imaginary numbers, `x*y + z` is not an imaginary number: suppose they are `x=ai` `y=bi` and `z=ci` where a,b,c are real numbers; then we get `ai*bi + ci` which simplifies to `-a*b + ci`, which is a complex number.

We could adjust it to define `proc fma(x: imag(64), y: imag(64), z: real(64))` (note that there are other variations we could include if we wanted), or we could just remove these `imag` overloads. I know I asked you to add these imag overloads my review. I'll leave it to you to choose one of these directions, and thanks.

Reply:
I'd be inclined to remove the overloads and wait until they are requested to add them again. Are we free to remove them, or should we go through the process of deprecating them first?

Reply:
IMO we can just remove them because they are unstable (and unlikely to be used). However deprecating them is also very easy.